src/txn/txn.c

```diff
@@ -1609,6 +1611,79 @@ __txn_mod_compare(const void *a, const void *b)
     return (aopt->u.op_col.recno < bopt->u.op_col.recno);
 }
 
+/*
+ * __wt_txn_op_list_clear_weak_hazard --
+ *     Walk the transaction op list clearing all the weak hazard pointers.
+ */
+int
+__wt_txn_op_list_clear_weak_hazard(WT_SESSION_IMPL *session)
+{
+    WT_DECL_RET;
+    WT_TXN *txn;
+    WT_TXN_OP *op;
+    u_int i;
+
+    txn = session->txn;
+    for (i = 0, op = txn->mod; i < txn->mod_count; i++, op++) {
+        if (op->whp != NULL)
+            WT_WITH_BTREE(session, op->btree, ret = __wt_hazard_weak_clear(session, &op->whp));
+        WT_RET(ret);
+    }
+    return (0);
+}
+
+/*
+ * __txn_resolve_weak_hazard_updates --
+ *     Resolve an uncommitted op if needed. Fetches the latest update pointer to be used as part of
+ *     the transaction operation.
+ */
+static int
+__txn_resolve_weak_hazard_updates(
+  WT_SESSION_IMPL *session, WT_TXN_OP *op, WT_CURSOR **cursorp, WT_REF **refp)
+{
+    WT_DECL_RET;
+    WT_REF *ref;
+    WT_TXN *txn;
+    WT_UPDATE *upd;
+
+    *refp = NULL;
+
+    ref = NULL;
+    txn = session->txn;
+
+    /* If we disabled resolution for some reason, or these are unsupported cases - do nothing. */
+    if (!txn->resolve_weak_hazard_updates)
+        return (0);
+    if (WT_IS_METADATA(op->btree->dhandle) ||
+      (op->type != WT_TXN_OP_BASIC_COL && op->type != WT_TXN_OP_INMEM_COL &&
+        op->type != WT_TXN_OP_BASIC_ROW && op->type != WT_TXN_OP_INMEM_ROW))
+        return (0);
+
+    /*
+     * Try to upgrade the weak pointer. If successful the page is still in memory and we are holding
+     * a strong hazard pointer. The caller will have to clear the strong hazard pointer when it is
+     * done using the page.
+     */
+    WT_WITH_BTREE(session, op->btree, ret = __wt_hazard_weak_upgrade(session, &op->whp, &ref));
+    if (ret == 0) {
+        *refp = ref;
+        return (0);
+    }
+
+    /* We could not upgrade the weak pointer - need to resolve the slow path. */
+    if (ret == EBUSY) {
+        WT_SAVE_DHANDLE(session, ret = __txn_search_uncommitted_op(session, op, cursorp, &upd));
+        WT_RET(ret);
+
+        /* Since we are not evicting yet - we expect to find the update we already have. */
```

Comment:
Can be worded better:

"Since we are not evicting yet and we don't resolve updates on keys with multiple updates we expect to find the update we already have."